### PR TITLE
initial project structure; share decorator

### PR DESCRIPTION
**Pull request by DmitryEfimenko**
In this commit:
* initial project structure
* configured the project to use Jest for unit-tests
* added `@Share` decorator
* added unit-tests for `@Share` decorator
* initial project description in the `README.md`

### CODE REVIEW

File: projects/lib/package.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "@ngneat/rxjs",
+  "version": "0.0.1",
+  "peerDependencies": {
+    "rxjs": "~6.4.0"
```
Comment (LayZeeDK) [inline projects/lib/package.json]: It should work with RxJS 6.4, 6.5 and so on, right? So we should use `"rxjs": "^6.4.0"` to not give consumers errors when using for example RxJS 6.5.

Comment (DmitryEfimenko) [reply projects/lib/package.json]: Good catch! I'll update when get home.

File: angular.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
+  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
+  "version": 1,
+  "newProjectRoot": "projects",
+  "projects": {
+    "lib": {
```
Comment (itayod) [inline angular.json]: add `@ngneat` scope

File: projects/lib/decorators/ng-package.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "lib": {
```
Comment (itayod) [inline projects/lib/decorators/ng-package.json]: isn't it missing a `"dist"` property?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+class BusinessLogic {
+  constructor(private api: MyApi) {
+    api.makeCall(1, 2).subscribe();
+    api.makeCall(1, 2).subscribe();
+    api.makeCall(1, 3).subscribe();
+  }
+}
+```
+
+You can provide the `when` option to specify the condition when the observable should be shared:
```
Comment (itayod) [inline README.md]: could you farther explain the use-case here?

File: projects/lib/decorators/share/share.spec.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      });
+
+      describe('GIVEN: when condition returns false', () => {
+        Given(() => {
+          when = (arg1: string, arg2: number) => {
+            return arg1 === '1' && arg2 === 2;
+          };
+        });
+
+        Then('Observable should have been called twise', () => {
```
Comment (itayod) [inline projects/lib/decorators/share/share.spec.ts]: twice*

File: projects/lib/decorators/share/share.spec.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      });
+    });
+
+    describe('GIVEN: two subs with different args', () => {
+      Given(() => {
+        setCall1Args('1', 1);
+        setCall2Args('2', 1);
+      });
+
+      Then('Observable should have been called twise', () => {
```
Comment (itayod) [inline projects/lib/decorators/share/share.spec.ts]: same

File: README.md (hunk 4 above)
Comment (itayod) [reply README.md]: also, mention that the context is the same context of the instance and the same signature of the method

File: projects/lib/decorators/share/share.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { Observable } from 'rxjs';
+import { share, tap } from 'rxjs/operators';
+
+export type MethodReturning<T> = (...args: any[]) => T;
+
+export interface ShareOptions {
+  when?: MethodReturning<boolean>;
+}
+
+export function Share(opts?: ShareOptions) {
```
Comment (itayod) [inline projects/lib/decorators/share/share.ts]:
```suggestion
export function Share(opts?: ShareOptions = {}) {
```

File: projects/lib/decorators/share/share.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      const args = arguments;
+      const key = JSON.stringify([...args]);
+
+      const updatedCall = (originalMethod.apply(context, args as any) as Observable<any>)
+        .pipe(
+          share(),
+          tap(() => cache.delete(key))
+        );
+
+      if ((!opts || !opts.when) || (opts.when as any).apply(context, args)) {
```
Comment (itayod) [inline projects/lib/decorators/share/share.ts]:
```suggestion
      if ((isFunction(opts.when)) && (opts.when as any).apply(context, args)) {
```

File: projects/lib/src/public-api.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+// primary entry-point is empty. All utilities should  be imported through their
+// individual entry-points. This file is needed to satisfy the "ng-packagr"
+// build which also requires a primary entry-point.
+
+export { };
```
[the remark this answers is not in the document]
Comment (DmitryEfimenko) [reply projects/lib/src/public-api.ts]: I purposefully didn't do that in order to mimic exports of the actual `rxjs` library. I foresee people will be using imports like this:
```ts
import { ... } from '@ngneat/rxjs/decorators';
import { ... } from '@ngneat/rxjs/operators';
```
This way imports seem cleaner to me.

Also, I'm unsure about the tree-shaking. Would re-exporting everything here affect tree-shaking in any negative way?

What do you think?

File: README.md (hunk 4 above)
Comment (DmitryEfimenko) [reply README.md]: Will do

Comment (DmitryEfimenko) [reply README.md]: Updated. Please let me know if you think anything else needs to be added.

**Comment by itayod** [comment]
@DmitryEfimenko seems like I did not understand what the decorator is doing before.
What are the use cases of it? seems like it solves pretty rare cases

**Comment by DmitryEfimenko** [comment]
@itayod the decorator does a very similar thing to the `share` rxjs operator except that it works with method calls that potentially accept arguments.

The most simple demonstration of the functionality is presented in the README file. In example 1 there are three method calls one after another. However, two of them are using the same arguments. Without the `@Share()` operator it would make 3 http calls. With it - it will make only two because there are only two unique sets of function arguments.

There are many articles out there explaining how to avoid duplicate http calls in Angular due to misunderstanding how rxjs works. [Here's one such article](https://blog.novanet.no/angular-pitfall-multiple-http-requests-with-rxjs-and-observable-async/). The problem comes down to having multiple `async` pipes that are subscribing to an observable that is an http request. The author solves the issue by adding `share` operator. But what if he had that http observable returned by a service method that had arguments? Simple `share` operator might not have fixed the issue because the intention is that calling the function with different arguments will not likely return the same data. There should be a separate http call made for each unique set of the function arguments. The `@Share` decorator takes that into account.

Please let me know if that makes sense. If that makes sense, perhaps I should copy-paste the above paragraphs into the README.

**Comment by itayod** [comment]
> But what if he had that http observable returned by a service method that had arguments?

It seems like a quite rare use-case to me.

Also, I am not seeing any advantage of using a Decorator instead of simply use the `share` operator, unless I am missing here something...

**Comment by DmitryEfimenko** [comment]
I think it addresses a very common use-case actually - avoiding duplicate http requests. However, it also handles properly a rather rare edge-cases.

**Pull request closed without merging**